tools/ecloud-revision-cleaner-tool/src/main/java/eu/europeana/cloud/jobs/RevisionRemoverJob.java

```diff
@@ -8,73 +8,154 @@
-                for (CloudTagsResponse cloudTagsResponse : cloudTagsResponses) {
-                    List<Representation> representations = recordServiceClient.getRepresentationsByRevision(cloudTagsResponse.getCloudId(), revisionInformation.getRepresentationName(), revisionInformation.getRevisionName(), revisionInformation.getRevisionProvider(), revisionInformation.getRevisionTimeStamp());
-                    for (Representation representation : representations) {
-                        if (representation.getRevisions().size() == 1) {
-                            removeRepresentationWithFilesAndRevisions(cloudTagsResponse, representation);
-                            break;
-                        } else {
-                            removeRevisionsOnly(representation);
-                        }
-                    }
+            } catch (MCSException | RuntimeException e) {
+                logConnectionException(e);
+            }
+
+            // If we have not obtained any results, we are done.
+            if (resultSlice == null || resultSlice.getResults() == null) {
+                executorService.shutdownNow();
+                break;
+            }
 
+            // Submit each response to the executor service. As soon as an exception occurs, we stop processing.
+            try {
+                for (CloudTagsResponse response : resultSlice.getResults()) {
+                    if (!executorService.isShutdown()) {
+                        executorService.submit(() -> removeRevision(response, executorService::shutdownNow));
+                    }
                 }
-                startFrom = resultSlice.getNextSlice();
-                LOGGER.info("The removal of revision " + revisionInformation.getRevisionName() + "_" + revisionInformation.getRevisionProvider() + "_" + revisionInformation.getRevisionTimeStamp() +
-                        " inside dataset: " + revisionInformation.getDataSet() + "_" + revisionInformation.getProviderId() + " is in progress!. This is the batch number: " + batchNumber);
-                batchNumber++;
+            } catch (RejectedExecutionException e) {
+                LOGGER.info("Could not submit task: probably the thread pool was closed due to an earlier problem. ", e);
+                break;
             }
-            while (startFrom != null);
+
+            // Go to next page.
+            startFrom = resultSlice.getNextSlice();
+            totalCounter += resultSlice.getResults().size();
+        }
+        while (startFrom != null && !executorService.isShutdown());
+        total.set(totalCounter);
+
+        // Trigger an orderly shutdown and wait for all tasks to finish.
+        final boolean success = !executorService.isShutdown();
+        executorService.shutdown();
+        try {
+            executorService.awaitTermination(100, TimeUnit.DAYS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```

Comment:
what is a purpose of doing this?
is only sets interrupted flag in thread and program flow go further?

Reply:
This call sets the interrupted flag, because it is not set anymore if there is an interrupted exception. 

I continued the flow because this thread is just waiting for it's children to complete.

But we can also choose to interrupt the executorService and then stop the program flow. This would be a very good alternative, actually. If you'd like to change it, please go for it.